python/pyarrow/lib_gpu.pyx

```diff
@@ -0,0 +1,809 @@
+
+    To create a CudaHostBuffer instance, use
+
+      <CudaDeviceManager instance>.allocate_host(<nbytes>)
+
+    To free memory allocated in CudaHostBuffer instance, use
+
+      <CudaDeviceManager instance>.free_host(<CudaHostBuffer instance>)
+
+    The memory is also automatically freed when CudaHostBuffer
+    instance is deleted.
+
+    Note: after freeing CudaHostBuffer memory, the instance should be
+    discarded as unusable.
+    """
+
+    def __cinit__(self):
+        self._freed = True
+
+    def __init__(self):
+        raise TypeError("Do not call CudaHostBuffer's constructor directly,"
+                        " use `allocate_cuda_host_buffer` function instead.")
+
+    cdef void init_host(self, const shared_ptr[CCudaHostBuffer] & buffer):
+        self.host_buffer = buffer
+        self.init(< shared_ptr[CBuffer] > buffer)
+        self._freed = False
+
+    @property
+    def size(self):
+        if self._freed:
+            return 0
+        return self.host_buffer.get().size()
+
+    def __dealloc__(self):
+        if not self._freed:
+            CudaDeviceManager().free_host(self)
+
+
+cdef class CudaBufferReader(NativeFile):
+    """File interface for zero-copy read from CUDA buffers.
+
+    Note: Read methods return pointers to device memory. This means
+    you must be careful using this interface with any Arrow code which
+    may expect to be able to do anything other than pointer arithmetic
+    on the returned buffers.
+    """
+    def __cinit__(self, CudaBuffer obj):
+        self.buffer = obj
+        self.reader = new CCudaBufferReader(self.buffer.buffer)
+        self.rd_file.reset(self.reader)
+        self.is_readable = True
+        self.closed = False
+
+    def read_buffer(self, nbytes=None):
+        """Read into new device buffer.
+
+        Parameters
+        ----------
+        nbytes : int, default None
+          Specify the number of bytes to read. Default: None (read all
+          remaining bytes).
+
+        Returns
+        -------
+        cbuf : CudaBuffer
+          New device buffer.
+        """
+        cdef:
+            int64_t c_nbytes
+            int64_t bytes_read = 0
+            shared_ptr[CCudaBuffer] output
+
+        if nbytes is None:
+            c_nbytes = self.size() - self.tell()
+        else:
+            c_nbytes = nbytes
+
+        with nogil:
+            check_status(self.reader.Read(c_nbytes,
+                                          < shared_ptr[CBuffer] * > & output))
+
+        return pyarrow_wrap_cudabuffer(output)
+
+
+cdef class CudaBufferWriter(NativeFile):
+    """File interface for writing to CUDA buffers, with optional buffering.
+    """
+    def __cinit__(self, CudaBuffer buffer):
+        self.buffer = buffer
+        self.writer = new CCudaBufferWriter(self.buffer.cuda_buffer)
+        self.wr_file.reset(self.writer)
+        self.is_writable = True
+        self.closed = False
+
+    def writeat(self, int64_t position, object data):
```

Comment:
May want to add this to `NativeFile` for writable files that support seek(). Feel free to create followup JIRA

Reply:
ARROW-3220